torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -0,0 +1,270 @@
+    }
+
+    // Add the original graphRoot as an edge.
+    edges.emplace_back(graphRoot, 0);
+
+    // Create a dummy GraphRoot and run init_to_execute with it.
+    GraphRoot dummyRoot(edges, {});
+    graphTask->init_to_execute(dummyRoot, outputEdges);
+
+    // Mark all 'RecvRPCBackward' as needing execution.
+    for (const auto& recvBackwardEdge : recvBackwardEdges) {
+      graphTask->exec_info_[recvBackwardEdge.function.get()].needed_ = true;
+    }
+  }
+
+  // Let autograd context take ownership of the GraphTask.
+  autogradContext.setGraphTask(std::move(graphTask));
+}
+
+void DistEngine::runEngineAndAccumulateGradients(
+    DistAutogradContext& autogradContext,
+    const std::shared_ptr<Node>& graphRoot,
+    const edge_list& outputEdges) {
+  // Kick off autograd computation with the root node and retrieve all the
+  // gradients.
+  // TODO: make this non-blocking
+  // (https://github.com/pytorch/pytorch/issues/26359)
+  variable_list grads = engine_.execute_with_graph_task(
+      autogradContext.retrieveGraphTask(), graphRoot);
+
+  // Accumulate all the gradients in the context.
+  TORCH_INTERNAL_ASSERT(grads.size() == outputEdges.size());
+  for (size_t i = 0; i < grads.size(); i++) {
+    // It is possible that the grad is not defined since a separate invocation
+    // of the autograd engine on the same node might actually compute this
+    // gradient.
+    // Also accumulate grads only for AccumulateGrad function.
+    if (grads[i].defined() &&
+        dynamic_cast<AccumulateGrad*>(outputEdges[i].function.get())) {
+      auto& variable =
+          std::static_pointer_cast<AccumulateGrad>(outputEdges[i].function)
+              ->variable;
+      autogradContext.accumulateGrad(variable, grads[i]);
+    }
+  }
+}
+
+void DistEngine::executeSendFunction(
+    DistAutogradContext& autogradContext,
+    const std::shared_ptr<Node>& sendFunction) {
+  std::unique_lock<std::mutex> lock(initializedContextIdsLock_);
+  if (initializedContextIds_.find(autogradContext.contextId()) ==
+      initializedContextIds_.end()) {
+    edge_list outputEdges;
+    // Pass in a dummy graphRoot since all send functions are the roots.
+    auto dummyRoot = std::make_shared<GraphRoot>(edge_list(), variable_list());
+    computeDependencies(autogradContext, {}, {}, dummyRoot, outputEdges);
+
+    // Mark the autograd context id as initialized and unlock.
+    initializedContextIds_.insert(autogradContext.contextId());
+    lock.unlock();
+
+    // Enqueue the current send function.
+    auto& graphTask = autogradContext.retrieveGraphTask();
+    engine_.enqueue_blocked_task_on_cpu(torch::autograd::NodeTask(
+        &graphTask, sendFunction, torch::autograd::InputBuffer(0)));
+
+    // Run the autograd engine.
+    runEngineAndAccumulateGradients(autogradContext, dummyRoot, outputEdges);
+
+    // Wait for all of the outstanding rpcs to complete.
+    autogradContext.clearAndWaitForOutStandingRpcs();
```

Comment:
This will block one RPC thread, right? This is fine on trainers but not so on the ps. Do we have any plan to make this non blocking? I recall we once had a discussion to let autograd engine expose a future.

Reply:
Yes, we probably need to make this non-blocking as well, currently the `runEngine` above will already block the thread.